### PR TITLE
[mlir][spirv] Implement vector unrolling for `convert-to-spirv` pass

### PR DESCRIPTION
**Pull request by obtuseangleAZ**
### Description
This PR builds on #99872. It implements a minimal version of function body vector unrolling to convert vector types into 1D and with a size supported by SPIR-V (2, 3 or 4 depending on the original dimension). The ops that are currently supported include those with elementwise traits (e.g. `arith.addi`), `vector.reduction` and `vector.transpose`. This PR also includes new LIT tests that only check for vector unrolling.

### Future Plans
- Support more ops

### CODE REVIEW

File: mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp [hunk 1]
```diff
@@ -56,6 +58,78 @@ struct ConvertToSPIRVPass final
         return signalPassFailure();
     }
 
+    if (runVectorUnrolling) {
+
```
Comment (kuhar) [inline mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp:62]:
```suggestion
```

File: mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp [hunk 2]
```diff
@@ -56,6 +58,78 @@ struct ConvertToSPIRVPass final
...
+        if (failed(applyPatternsAndFoldGreedily(op, std::move(patterns)))) {
+          return signalPassFailure();
+        }
+      }
+
+      // Unroll vectors to native vector size.
+      {
+        RewritePatternSet patterns(context);
+        auto options = vector::UnrollVectorOptions().setNativeShapeFn(
+            [=](auto op) { return mlir::spirv::getNativeVectorShape(op); });
```
Comment (kuhar) [inline mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp:76]: What do we capture here?

File: mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp [hunk 3]
```diff
@@ -56,6 +58,78 @@ struct ConvertToSPIRVPass final
...
+        vector::BroadcastOp::getCanonicalizationPatterns(patterns, context);
+        vector::ShapeCastOp::getCanonicalizationPatterns(patterns, context);
+
+        if (failed(applyPatternsAndFoldGreedily(op, std::move(patterns))))
+          return signalPassFailure();
+      }
+
+      // Run all sorts of canonicalization patterns to clean up again.
+      {
+        RewritePatternSet patterns(context);
```
Comment (kuhar) [inline mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp:122]: Why do we run these separately from the patterns above?

File: mlir/include/mlir/Dialect/SPIRV/Transforms/SPIRVConversion.h [hunk 4]
```diff
@@ -189,6 +189,17 @@ Value getVulkanElementPtr(const SPIRVTypeConverter &typeConverter,
                           MemRefType baseType, Value basePtr,
                           ValueRange indices, Location loc, OpBuilder &builder);
 
+int getComputeVectorSize(int64_t size);
```
Comment (kuhar) [inline mlir/include/mlir/Dialect/SPIRV/Transforms/SPIRVConversion.h:192]: Now that this is a public function, could you add some documentation?

File: mlir/test/Conversion/ConvertToSPIRV/vector-unroll.mlir [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  // CHECK: %[[EXTRACT3:.*]] = vector.extract %[[ARG1]][1] : i32 from vector<3xi32>
+  // CHECK: %[[INSERT3:.*]] = vector.insert %[[EXTRACT3]], %[[INSERT2]] [1] : i32 into vector<2xi32>
+  // CHECK: %[[EXTRACT4:.*]] = vector.extract %[[ARG0]][2] : i32 from vector<3xi32>
+  // CHECK: %[[INSERT4:.*]] = vector.insert %[[EXTRACT4]], %[[CST]] [0] : i32 into vector<2xi32>
+  // CHECK: %[[EXTRACT5:.*]] = vector.extract %[[ARG1]][2] : i32 from vector<3xi32>
+  // CHECK: %[[INSERT5:.*]] = vector.insert %[[EXTRACT5]], %[[INSERT4]] [1] : i32 into vector<2xi32>
+  // CHECK: return %[[INSERT1]], %[[INSERT3]], %[[INSERT5]] : vector<2xi32>, vector<2xi32>, vector<2xi32>
+  %0 = vector.transpose %arg0, [1, 0] : vector<2x3xi32> to vector<3x2xi32>
+  return %0 : vector<3x2xi32>
+}
```
Comment (kuhar) [inline mlir/test/Conversion/ConvertToSPIRV/vector-unroll.mlir:102]: nit: missing newline

File: mlir/test/lib/Conversion/ConvertToSPIRV/TestSPIRVVectorUnrolling.cpp [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<spirv::SPIRVDialect, vector::VectorDialect>();
+  }
+
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    Operation *op = getOperation();
+
+    // Unroll vectors in function signatures to native vector size.
```
Comment (kuhar) [inline mlir/test/lib/Conversion/ConvertToSPIRV/TestSPIRVVectorUnrolling.cpp:40]: This is quite a lot of code, I'd consider moving all of these to a helper function(s) that both the test and the main pass can call.

**Review by kuhar — CHANGES_REQUESTED** [review]

File: mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp (hunk 2 above)
Comment (angelz913) [reply mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp:75]: This is the implementation in IREE. I think `[=]` captures all external variables by value, although in this case it seems we do not need to capture anything.

Comment (kuhar) [reply mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp:75]: Yeah so let's remove this if it's not needed

Comment (angelz913) [reply mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp]: Removed

File: mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp (hunk 3 above)
Comment (angelz913) [reply mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp]: These patterns were found to be redundant and have been removed.

File: mlir/include/mlir/Dialect/SPIRV/Transforms/SPIRVConversion.h [hunk 7]
```diff
@@ -202,6 +203,12 @@ SmallVector<int64_t> getNativeVectorShapeImpl(vector::TransposeOp op);
 // For general ops.
 std::optional<SmallVector<int64_t>> getNativeVectorShape(Operation *op);
 
+// Unroll vectors in function signatures to native size.
+bool unrollVectorsInSignatures(Operation *op, MLIRContext *context);
+
+// Unroll vectors in function bodies to native size.
+bool unrollVectorsInFuncBodies(Operation *op, MLIRContext *context);
```
Comment (kuhar) [inline mlir/include/mlir/Dialect/SPIRV/Transforms/SPIRVConversion.h:210]: In this form it's impossible to tell if `true` indicates success or failure. Return `LogicalResult` instead.

Comment (kuhar) [reply mlir/include/mlir/Dialect/SPIRV/Transforms/SPIRVConversion.h:210]: Also you can extract context from the op.

File: mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp [hunk 8]
```diff
@@ -1285,6 +1287,115 @@ Value mlir::spirv::getElementPtr(const SPIRVTypeConverter &typeConverter,
...
+  MLIRContext *context = op->getContext();
+
+  // Unroll vectors in function bodies to native vector size.
+  {
+    RewritePatternSet patterns(context);
+    auto options = vector::UnrollVectorOptions().setNativeShapeFn(
+        [](auto op) { return mlir::spirv::getNativeVectorShape(op); });
+    populateVectorUnrollPatterns(patterns, options);
+    if (failed(applyPatternsAndFoldGreedily(op, std::move(patterns))))
+      return llvm::failure();
```
Comment (kuhar) [inline mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp:1357]:
```suggestion
      return failure();
```

similar with `success` and `failure` elsewhere in this file

**Review by kuhar — APPROVED** [review]
LGTM % nit

File: mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp [hunk 9]
```diff
@@ -1285,6 +1287,115 @@ Value mlir::spirv::getElementPtr(const SPIRVTypeConverter &typeConverter,
...
+  SmallVector<int64_t> nativeSize(vectorType.getRank(), 1);
+  nativeSize.back() =
+      mlir::spirv::getComputeVectorSize(vectorType.getShape().back());
+  return nativeSize;
+}
+
+std::optional<SmallVector<int64_t>>
+mlir::spirv::getNativeVectorShape(Operation *op) {
+  if (OpTrait::hasElementwiseMappableTraits(op) && op->getNumResults() == 1) {
+    if (auto vecType = llvm::dyn_cast<VectorType>(op->getResultTypes()[0])) {
```
Comment (Hardcode84) [inline mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp:1323]: nit: drop `llvm::`

File: mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp [hunk 10]
```diff
@@ -1285,6 +1287,115 @@ Value mlir::spirv::getElementPtr(const SPIRVTypeConverter &typeConverter,
...
+      .Default([](Operation *) { return std::nullopt; });
+}
+
+LogicalResult mlir::spirv::unrollVectorsInSignatures(Operation *op) {
+  MLIRContext *context = op->getContext();
+  RewritePatternSet patterns(context);
+  populateFuncOpVectorRewritePatterns(patterns);
+  populateReturnOpVectorRewritePatterns(patterns);
+  GreedyRewriteConfig config;
+  config.strictMode = GreedyRewriteStrictness::ExistingOps;
```
Comment (Hardcode84) [inline mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp:1343]: Please add comment why we need `config.strictMode`

File: mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp [hunk 11]
```diff
@@ -1285,6 +1287,115 @@ Value mlir::spirv::getElementPtr(const SPIRVTypeConverter &typeConverter,
...
+    vector::InsertOp::getCanonicalizationPatterns(patterns, context);
+    vector::ExtractOp::getCanonicalizationPatterns(patterns, context);
+
+    // Trimming leading unit dims may generate broadcast/shape_cast ops. Clean
+    // them up.
+    vector::BroadcastOp::getCanonicalizationPatterns(patterns, context);
+    vector::ShapeCastOp::getCanonicalizationPatterns(patterns, context);
+
+    if (failed(applyPatternsAndFoldGreedily(op, std::move(patterns))))
+      return llvm::failure();
```
Comment (Hardcode84) [inline mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp:1394]: drop `llvm::`

File: mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp [hunk 12]
```diff
@@ -1098,13 +1094,19 @@ struct ReturnOpVectorUnroll final : OpRewritePattern<func::ReturnOp> {
...
+      SmallVector<int64_t> extractShape(originalShape.size(), 1);
+      extractShape.back() = targetShape->back();
       SmallVector<Type> newTypes;
       Value returnValue = returnOp.getOperand(origResultNo);
       for (SmallVector<int64_t> offsets :
            StaticTileOffsetRange(originalShape, *targetShape)) {
         Value result = rewriter.create<vector::ExtractStridedSliceOp>(
-            loc, returnValue, offsets, *targetShape, strides);
+            loc, returnValue, offsets, extractShape, strides);
+        SmallVector<int64_t> extractIndices(originalShape.size() - 1, 0);
```
Comment (Hardcode84) [inline mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp:1106]: Move `extractIndices` inside the `if`

File: mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp (hunk 11 above)
Comment (angelz913) [reply mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp]: Dropped

**Review by Hardcode84 — APPROVED** [review]

**Pull request merged**